Parsedown.php

```diff
@@ -1485,10 +1503,80 @@ function parse($text)
         return $markup;
     }
 
+    protected function sanitiseElement(array $Element)
+    {
+        static $badAttributeChars = "\"'= \t\n\r\0\x0B";
+        static $safeUrlNameToAtt  = array(
+            'a'   => 'href',
+            'img' => 'src',
+        );
+
+        if (isset($safeUrlNameToAtt[$Element['name']]))
+        {
+            $Element = $this->filterUnsafeUrlInAttribute($Element, $safeUrlNameToAtt[$Element['name']]);
+        }
+
+        if ( ! empty($Element['attributes']))
+        {
+            foreach ($Element['attributes'] as $att => $val)
+            {
+                # clear out nulls
+                if ($val === null)
+                {
+                    unset($Element['attributes'][$att]);
+                }
+                # filter out badly parsed attribute
+                elseif (strpbrk($att, $badAttributeChars) !== false)
+                {
+                    unset($Element['attributes'][$att]);
+                }
+            }
+
+            $onEventAttributeKeys = preg_grep('/^on/i', array_keys($Element['attributes']));
+
+            foreach ($onEventAttributeKeys as $att)
+            {
+                unset($Element['attributes'][$att]);
+            }
+        }
+
+        return $Element;
+    }
+
+    protected function filterUnsafeUrlInAttribute(array $Element, $attribute)
+    {
+        if ($this->safeLinksEnabled)
+        {
+            $safe = false;
+
+            foreach ($this->safeLinksWhitelist as $scheme)
+            {
+                if (stripos($Element['attributes'][$attribute], $scheme) === 0)
```

Comment:
Even if I could enumerate all the protocols that allowed scripting, what if a new one comes into existence?

I started trying to write this to give us a bigger whitelist

```bash
curl -s https://www.iana.org/assignments/uri-schemes/uri-schemes.txt | egrep 'Permanent|Provisional|Historical' | cut -d ' ' -f 1 | sed "s/^/'/" | sed "s/$/:',/"
```

But `data:` is listed in there (which makes sense I suppose).

So it then changed to this (addition of `egrep -v '^data|^javascript|^$'`)

```bash
curl -s https://www.iana.org/assignments/uri-schemes/uri-schemes.txt | egrep 'Permanent|Provisional|Historical' | cut -d ' ' -f 1 | egrep -v '^data|^javascript|^$'  | sed "s/^/'/" | sed "s/$/:',/"
```

But really, I'm just moving the blacklist somewhere else by doing this. There's no way I can honestly go through all those protocols and decide whether or not they would permit scripting in certain situations.

I think the only thing we can sensibly do is start with some common protocols on the whitelist, and then add new ones after a bug report is issued, and someone evaluates whether adding that protocol would permit scripting.